src/platform/packages/shared/kbn-scout/src/cli/config_discovery.ts

```diff
@@ -7,92 +7,230 @@
+    );
+  });
+};
 
-    const runnablePluginCount = getCountByType(runnableConfigs, 'plugin');
-    const runnablePackageCount = getCountByType(runnableConfigs, 'package');
+const handleFlattenedOutput = (
+  filteredModules: ModuleDiscoveryInfo[],
+  flagsReader: FlagsReader,
+  log: ToolingLog
+): void => {
+  // Apply CI filtering if save flag is set (for consistency with non-flattened behavior)
+  const modulesToFlatten = flagsReader.boolean('save')
+    ? filterModulesByScoutCiConfig(log, filteredModules)
+    : filteredModules;
 
-    const dirPath = path.dirname(SCOUT_PLAYWRIGHT_CONFIGS_PATH);
+  const flattenedConfigs = flattenModulesByServerRunFlag(modulesToFlatten);
 
-    if (!fs.existsSync(dirPath)) {
-      fs.mkdirSync(dirPath, { recursive: true });
-    }
+  if (flagsReader.boolean('save')) {
+    saveFlattenedConfigGroups(flattenedConfigs, log);
+    return;
+  }
 
-    fs.writeFileSync(
-      SCOUT_PLAYWRIGHT_CONFIGS_PATH,
-      JSON.stringify(Object.fromEntries(runnableConfigs), null, 2)
-    );
+  logFlattenedConfigs(flattenedConfigs, log);
+};
+
+const handleNonFlattenedOutput = (
+  filteredModules: ModuleDiscoveryInfo[],
+  flagsReader: FlagsReader,
+  log: ToolingLog
+): void => {
+  if (flagsReader.boolean('save')) {
+    const filteredForCiModules = filterModulesByScoutCiConfig(log, filteredModules);
+    saveModuleDiscoveryInfo(filteredForCiModules, log);
+
+    const { plugins: savedPluginCount, packages: savedPackageCount } =
+      countModulesByType(filteredForCiModules);
 
     log.info(
-      `${finalMessage}.\nSaved ${runnablePluginCount} plugin(s) and ${runnablePackageCount} package(s) to '${SCOUT_PLAYWRIGHT_CONFIGS_PATH}'`
+      `Scout configs were filtered for CI. Saved ${savedPluginCount} plugin(s) and ${savedPackageCount} package(s) to '${SCOUT_PLAYWRIGHT_CONFIGS_PATH}'`
     );
-
     return;
   }
 
   if (flagsReader.boolean('validate')) {
-    validateWithScoutCiConfig(log, scoutConfigs);
+    filterModulesByScoutCiConfig(log, filteredModules);
+    return;
+  }
+
+  logDiscoveredModules(filteredModules, log);
+};
+
+// Discovers and processes Playwright configuration files with Scout tests
+export const runDiscoverPlaywrightConfigs = (flagsReader: FlagsReader, log: ToolingLog): void => {
+  const target = (flagsReader.enum('target', TARGET_TYPES) || 'all') as TargetType;
+  const targetTags = getTestTagsForTarget(target);
+  const flatten = flagsReader.boolean('flatten');
+
+  // Build initial module discovery info
+  const modulesWithTests = buildModuleDiscoveryInfo();
+  // Filter modules by target tags and compute server run flags
+  const filteredModules = filterModulesByTargetTags(modulesWithTests, targetTags);
+  // Handle output based on flatten flag
+  if (flatten) {
+    handleFlattenedOutput(filteredModules, flagsReader, log);
+  } else {
+    handleNonFlattenedOutput(filteredModules, flagsReader, log);
   }
```

Comment:
Why we need 2 formats: in regular Kibana CI we spin up a dedicated worker for each plugin/package with Scout tests, but for Cloud runs we would like to move forward and run configs from different plugins/package sequentially against the same project/deployment.

The general idea is to move away from "worker per module" approach in regular pipelines, but we will address it later with configs balancer.